worstcase/state.go

```diff
@@ -0,0 +1,266 @@
+	if err != nil {
+		return nil, err
+	}
+	return &State{
+		hooks:              hooks,
+		config:             config,
+		db:                 db,
+		clock:              settled.ExecutedByGasTime(),
+		expectedParentHash: settled.Hash(),
+	}, nil
+}
+
+const rateToMaxBlockSize = saeparams.Tau * saeparams.Lambda
+
+var (
+	errNonConsecutiveBlocks = errors.New("non-consecutive blocks")
+	errQueueFull            = errors.New("queue full")
+)
+
+// StartBlock updates the worst-case state to the beginning of the provided
+// block.
+//
+// It is not necessary for [types.Header.GasLimit] nor [types.Header.BaseFee] to
+// be set. However, all other fields should be populated and
+// [types.Header.ParentHash] must match the previous block's hash.
+//
+// If the queue is too full to accept another block, an error is returned.
+func (s *State) StartBlock(h *types.Header) error {
+	if s.expectedParentHash != h.ParentHash {
+		return fmt.Errorf("%w: expected parent hash of %s but was %s",
+			errNonConsecutiveBlocks,
+			s.expectedParentHash,
+			h.ParentHash,
+		)
+	}
+
+	s.clock.BeforeBlock(s.hooks, h)
+	s.blockSize = 0
+
+	const (
+		maxQSizeMultiplier = 2
+		// In order to avoid overflow when calculating the queue size, we cap
+		// the maximum gas rate to a safe value.
+		//
+		// This follows from:
+		//   maxBlockSize = maxRate * Tau * Lambda
+		//   maxQSizeInStart = maxQSizeMultiplier * maxBlockSize
+		//   maxQSizeInFinish = maxQSizeInStart + maxBlockSize
+		maxRate gas.Gas = math.MaxUint64 / rateToMaxBlockSize / (maxQSizeMultiplier + 1)
+	)
+	r := min(s.clock.Rate(), maxRate)
+	s.maxBlockSize = r * rateToMaxBlockSize
+	if maxQSize := maxQSizeMultiplier * s.maxBlockSize; s.qSize > maxQSize {
+		return fmt.Errorf("%w: current size %d exceeds maximum size %d", errQueueFull, s.qSize, maxQSize)
+	}
+
+	s.baseFee = s.clock.BaseFee()
+
+	// expectedParentHash is used prior to modifying the GasLimit and BaseFee to
+	// ensure that historical block hashes are not modified.
+	s.expectedParentHash = h.Hash()
+	s.curr = types.CopyHeader(h)
+	s.curr.GasLimit = uint64(s.maxBlockSize)
+	s.curr.BaseFee = s.baseFee.ToBig()
+
+	// For both rules and signer, we MUST use the block's timestamp, not the
+	// execution clock's, otherwise we might enable an upgrade too early.
+	s.rules = s.config.Rules(h.Number, true /*merge*/, h.Time)
+	s.signer = types.MakeSigner(s.config, h.Number, h.Time)
+	return nil
+}
+
+// GasLimit returns the available gas limit for the current block.
+func (s *State) GasLimit() uint64 {
+	return uint64(s.maxBlockSize)
+}
+
+// BaseFee returns the worst-case base fee for the current block.
+func (s *State) BaseFee() *uint256.Int {
+	return s.baseFee
+}
+
+type (
+	// AccountDebit includes an amount that an account should have debited,
+	// along with the nonce used to debit the account.
+	AccountDebit = hook.AccountDebit
+	// Op is an operation that can be applied to a [State].
+	Op = hook.Op
```

Comment:
I'm comparing this with https://github.com/ava-labs/libevm/blob/main/core/state_transition.go#L274, most of them are checked in `ValidateTransaction`. The ones I could not see are:
1- Do we need to check if the sender is eoa? (https://github.com/ava-labs/libevm/blob/main/core/state_transition.go#L293)
2- Should we check the msg.Value? https://github.com/ava-labs/libevm/blob/main/core/state_transition.go#L407-L414

Reply:
I ended up adding the EOA check.